src/Interpreters/HashJoin/HashJoinMethodsImpl.h

```diff
@@ -363,6 +395,26 @@ size_t HashJoinMethods<KIND, STRICTNESS, MapsTemplate>::joinRightColumns(
             }
         }
 
+        /// Do data prefetch
+        if constexpr (HasGetKeyHolderMemberFunc<KeyGetter, Arena>)
+        {
+            using KeyHolder = decltype(key_getter_vector[0].getKeyHolder(0, std::declval<Arena &>()));
+            auto & data_prefetch = *(mapv[0]);
+            if constexpr (HasPrefetchMemberFunc<decltype(data_prefetch), KeyHolder>)
+            {
+                if (i + prefetch_ahead_distance < rows && added_columns.join_on_keys.size() < prefetch_small_loop_size)
+                {
+                    const auto & join_keys_prefetch = added_columns.join_on_keys[0];
+                    if (!(join_keys_prefetch.null_map && (*join_keys_prefetch.null_map)[i + prefetch_ahead_distance]) &&
```

Comment:
usually such conditions make prefetch efficiency worse because there is a dependency: before you could prefetch `hash_table[i]` you need to first calculate some expression over `some_other_ds[i]` and if it is also not in cache we will have a stall on this access. so if we could get rid of this checks - it'd be nice.